src/core/fonts.js

```diff
@@ -2050,35 +2050,45 @@ var Font = (function FontClosure() {
         var oldGlyfData = glyf.data;
         var oldGlyfDataLength = oldGlyfData.length;
         var newGlyfData = new Uint8Array(oldGlyfDataLength);
-        var startOffset = itemDecode(locaData, 0);
-        var writeOffset = 0;
+
+        // The spec says the offsets should be in ascending order, however
+        // this is not true for some fonts or they use the offset of 0 to mark a
+        // glyph as missing. OTS requires the offsets in order and not to
+        // be zero, so we must sort and rebuild the loca table and potentially
+        // re-arrange the glyf data.
+        var i, j;
+        const locaEntries = [];
+        for (i = 0, j = 0; i <= numGlyphs; i++, j += itemSize) {
+          let offset = itemDecode(locaData, j);
+          if (offset > oldGlyfDataLength) {
+            offset = oldGlyfDataLength;
+          }
+          locaEntries.push({
+            index: i,
+            offset,
+            endOffset: 0,
+          });
+        }
+        locaEntries.sort((a, b) => {
+          return a.offset - b.offset;
+        });
+        // Now the offsets are sorted, calculate the end offset of each glyph.
+        for (i = 0; i < numGlyphs; i++) {
+          locaEntries[i].endOffset = locaEntries[i + 1].offset;
+        }
```

Comment:
Doesn't this mean that the *last* loca-table entry will still have a zero `endOffset`, i.e. `locaEntries[numGlyphs].endOffset === 0`; is that not a problem?

Unless I'm overlooking something here, this means that `sanitizeGlyph` will just reject the glyph since it's deemed empty!?

Reply:
It's a bit confusing since there are `numGlyphs + 1` loca entries. Down below we don't iterate over the last loca entry (`i < numGlyphs`), but it's data is effectively saved in the previous entry when endOffset is read here.

I think I'll add some more comments, and maybe change the above `i <= numGlyphs` to `i < numGlyphs + 1` to make it more explicit.